In [2]:
pip install torch torchvision torchaudio

Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 59.6 MB 36.9 MB/s eta 0:00:01
     |████████████████████████████████| 1.5 MB 45.4 MB/s eta 0:00:01
     |████████████████████████████████| 1.7 MB 3.1 MB/s eta 0:00:01
  Using cached fsspec-2023.12.0-py3-none-any.whl (168 kB)
  Using cached filelock-3.13.1-py3-none-any.whl (11 kB)
     |████████████████████████████████| 133 kB 3.1 MB/s eta 0:00:01
     |████████████████████████████████| 1.6 MB 44.3 MB/s eta 0:00:01
     |████████████████████████████████| 5.7 MB 49.5 MB/s eta 0:00:01
     |████████████████████████████████| 3.3 MB 57.7 MB/s eta 0:00:01
     |████████████████████████████████| 62 kB 6.9 MB/s  eta 0:00:01
     |████████████████████████████████| 14.0 MB 57.7 MB/s eta 0:00:01
     |████████████████████████████████| 120 kB 33.4 MB/s eta 0:00:01
     |████████████████████████████████| 104 kB 36.7 MB/s eta 0:00:01
     |████████████████████████████████| 61 kB 625 

In [3]:
pip install transformers

Defaulting to user installation because normal site-packages is not writeable
  Using cached transformers-4.35.2-py3-none-any.whl (7.9 MB)
     |████████████████████████████████| 426 kB 4.0 MB/s eta 0:00:01
     |████████████████████████████████| 78 kB 12.6 MB/s eta 0:00:01
  Using cached huggingface_hub-0.19.4-py3-none-any.whl (311 kB)
     |████████████████████████████████| 291 kB 40.8 MB/s eta 0:00:01
     |████████████████████████████████| 174 kB 37.5 MB/s eta 0:00:01
     |████████████████████████████████| 2.5 MB 40.7 MB/s eta 0:00:01
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [4]:
pip install --upgrade transformers

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [5]:
pip show transformers

Name: transformers
Version: 4.35.2
Summary: State-of-the-art Machine Learning for JAX, PyTorch and TensorFlow
Home-page: https://github.com/huggingface/transformers
Author: The Hugging Face team (past and future) with the help of all our contributors (https://github.com/huggingface/transformers/graphs/contributors)
Author-email: transformers@huggingface.co
License: Apache 2.0 License
Location: /Users/kariprimiano/Library/Python/3.9/lib/python/site-packages
Requires: tqdm, regex, tokenizers, pyyaml, numpy, safetensors, packaging, filelock, huggingface-hub, requests
Required-by: 
Note: you may need to restart the kernel to use updated packages.


In [6]:
pip install --upgrade s3fs

Defaulting to user installation because normal site-packages is not writeable
  Using cached s3fs-2023.12.0-py3-none-any.whl (29 kB)
  Using cached aiobotocore-2.8.0-py3-none-any.whl (75 kB)
     |████████████████████████████████| 387 kB 4.9 MB/s eta 0:00:01
  Using cached botocore-1.33.1-py3-none-any.whl (11.6 MB)
     |████████████████████████████████| 46 kB 13.4 MB/s eta 0:00:01
     |████████████████████████████████| 81 kB 26.0 MB/s eta 0:00:01
     |████████████████████████████████| 61 kB 17.1 MB/s eta 0:00:01
     |████████████████████████████████| 143 kB 38.2 MB/s eta 0:00:01
  Attempting uninstall: urllib3
    Found existing installation: urllib3 2.1.0
    Uninstalling urllib3-2.1.0:
      Successfully uninstalled urllib3-2.1.0
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [7]:
pip install fsspec==2023.3.0

Defaulting to user installation because normal site-packages is not writeable
  Using cached fsspec-2023.3.0-py3-none-any.whl (145 kB)
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2023.12.0
    Uninstalling fsspec-2023.12.0:
      Successfully uninstalled fsspec-2023.12.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
s3fs 2023.12.0 requires fsspec==2023.12.0, but you have fsspec 2023.3.0 which is incompatible.
huggingface-hub 0.19.4 requires fsspec>=2023.5.0, but you have fsspec 2023.3.0 which is incompatible.
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [8]:
pip install nltk

Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 1.5 MB 4.5 MB/s eta 0:00:01
  Using cached click-8.1.7-py3-none-any.whl (97 kB)
     |████████████████████████████████| 302 kB 44.8 MB/s eta 0:00:01
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [10]:
pip install pandas

Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 11.0 MB 4.1 MB/s eta 0:00:01
     |████████████████████████████████| 341 kB 40.4 MB/s eta 0:00:01
     |████████████████████████████████| 502 kB 52.2 MB/s eta 0:00:01
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [11]:
pip install numpy

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [13]:
pip install scikit-learn

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [8]:
from transformers import pipeline
import nltk
from nltk.corpus import stopwords
import string
import requests
import pandas as pd
import time
import json
import zipfile
import os
import re
import ast
import torch


import warnings
warnings.filterwarnings('ignore')

print(torch.__version__)


2.1.1


In [9]:
master_data = pd.read_csv('master_anime_data.csv')

In [13]:
# Sample a subset of your data for processing
sampled_data = master_data.sample(frac=0.1, random_state=13)

In [14]:
# Download nltk stop words
nltk.download('stopwords')

# Load the emotion classification pipeline
model_name = "j-hartmann/emotion-english-distilroberta-base"
emotion_classifier = pipeline('text-classification', model=model_name, return_all_scores=True)

# Set English stop words
stop_words = set(stopwords.words('english'))

# Function for advanced text cleaning
def clean_text(text):
    text = text.lower()
    text = re.sub(r'<[^>]+>', '', text)
    text = text.translate(str.maketrans('', '', string.punctuation))  # Remove punctuation
    text = ' '.join([word for word in text.split() if word not in stop_words])  # Remove stop words
    return text

# Function to classify emotions of a text
def classify_emotions(text):
    cleaned_text = clean_text(text)  # Clean the text using clean_text function
    predictions = emotion_classifier(cleaned_text)
    return {emotion['label']: emotion['score'] for emotion in predictions[0]}

# Apply the model to the 'synopsis' column
sampled_data['emotion_scores'] = sampled_data['synopsis'].astype(str).apply(classify_emotions)

# Inspect results
sampled_data[['synopsis', 'emotion_scores']].head()


[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/kariprimiano/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


synopsis  \
19101   After the accident in which she lost her mothe...   
217920  Naota Nandaba is an ordinary sixth grader livi...   
267563  Story begins as Momiji's chastity is taken by ...   
9909    Tanya Degurechaff is a young soldier infamous ...   
165782  Hachiman Hikigaya is an apathetic high school ...   

                                           emotion_scores  
19101   {'anger': 0.010413373820483685, 'disgust': 0.0...  
217920  {'anger': 0.18671022355556488, 'disgust': 0.00...  
267563  {'anger': 0.007550527341663837, 'disgust': 0.0...  
9909    {'anger': 0.5558895468711853, 'disgust': 0.003...  
165782  {'anger': 0.011294235475361347, 'disgust': 0.0...

In [15]:
sampled_data['emotion_scores'].value_counts

<bound method IndexOpsMixin.value_counts of 19101     {'anger': 0.010413373820483685, 'disgust': 0.0...
217920    {'anger': 0.18671022355556488, 'disgust': 0.00...
267563    {'anger': 0.007550527341663837, 'disgust': 0.0...
9909      {'anger': 0.5558895468711853, 'disgust': 0.003...
165782    {'anger': 0.011294235475361347, 'disgust': 0.0...
                                ...                        
182181    {'anger': 0.0200864989310503, 'disgust': 0.000...
57783     {'anger': 0.007323819678276777, 'disgust': 0.0...
154919    {'anger': 0.005832878407090902, 'disgust': 0.0...
265514    {'anger': 0.004111292771995068, 'disgust': 0.0...
85081     {'anger': 0.022216347977519035, 'disgust': 0.0...
Name: emotion_scores, Length: 34988, dtype: object>

### Expanding 'emotion_scores' into separate columns

In [16]:
# Check if the emotion_scores column needs to be parsed from string to dictionary
if isinstance(sampled_data['emotion_scores'].iloc[0], str):
    sampled_data['emotion_scores'] = sampled_data['emotion_scores'].apply(ast.literal_eval)

# Expand the emotion_scores column into separate columns
emotion_scores = sampled_data['emotion_scores'].apply(pd.Series)

# Join the expanded emotion scores dataframe with the original data
data_with_emotions = sampled_data.join(emotion_scores)

# Display the first few rows to verify the expansion
data_with_emotions.head()


uid_review           profile  anime_uid  \
19101        27270           Elfsire        120   
217920      228044      merryfistmas        227   
267563       21744         ravenrage       3303   
9909        248039  Tyrannicswine117      32615   
165782      215510       ThePunzalan      14813   

                                                   scores  Overall  Story  \
19101   {'Overall': '8', 'Story': '8', 'Animation': '1...        8      8   
217920  {'Overall': '10', 'Story': '10', 'Animation': ...       10     10   
267563  {'Overall': '6', 'Story': '3', 'Animation': '8...        6      3   
9909    {'Overall': '8', 'Story': '9', 'Animation': '8...        8      9   
165782  {'Overall': '9', 'Story': '9', 'Animation': '9...        9      9   

        Animation  Sound  Character  Enjoyment  ...  \
19101          10      8         10          7  ...   
217920         10     10         10         10  ...   
267563          8      5          6          7  ...   
9909            8      8          9          8  ...   
165782          9      9         10          9  ...   

                                          favorites_anime  \
19101   ['14713', '14289', '4081', '14467', '16498', '...   
217920  ['18679', '22135', '227', '23847', '5114', '11...   
267563                                            ['134']   
9909    ['5114', '4181', '9253', '18679', '1', '2001',...   
165782  ['14813', '30831', '15809', '4224', '31715', '...   

                                                  link_y  \
19101            https://myanimelist.net/profile/Elfsire   
217920      https://myanimelist.net/profile/merryfistmas   
267563         https://myanimelist.net/profile/ravenrage   
9909    https://myanimelist.net/profile/Tyrannicswine117   
165782       https://myanimelist.net/profile/ThePunzalan   

                                           emotion_scores     anger   disgust  \
19101   {'anger': 0.010413373820483685, 'disgust': 0.0...  0.010413  0.000746   
217920  {'anger': 0.18671022355556488, 'disgust': 0.00...  0.186710  0.002664   
267563  {'anger': 0.007550527341663837, 'disgust': 0.0...  0.007551  0.001111   
9909    {'anger': 0.5558895468711853, 'disgust': 0.003...  0.555890  0.003051   
165782  {'anger': 0.011294235475361347, 'disgust': 0.0...  0.011294  0.000683   

            fear       joy   neutral   sadness  surprise  
19101   0.197999  0.028194  0.013339  0.689401  0.059908  
217920  0.069133  0.091217  0.139144  0.045635  0.465496  
267563  0.001951  0.658758  0.077451  0.017352  0.235827  
9909    0.173663  0.078683  0.091056  0.014780  0.082877  
165782  0.000702  0.938779  0.007481  0.031252  0.009808  

[5 rows x 34 columns]

## Classification Model for Mood Categorization

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Define the mood categories based on the highest emotion score
mood_categories = data_with_emotions[['anger', 'disgust', 'fear', 'joy', 'neutral', 'sadness', 'surprise']].idxmax(axis=1)

# Add the mood categories to the DataFrame
data_with_emotions['mood_category'] = mood_categories

# Prepare the data
X = data_with_emotions[['anger', 'disgust', 'fear', 'joy', 'neutral', 'sadness', 'surprise']]  # Features
y = data_with_emotions['mood_category']  # Labels

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the classifier
classifier = RandomForestClassifier(random_state=42)

# Train the classifier
classifier.fit(X_train, y_train)

# Make predictions
y_pred = classifier.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')


### Rule-Based Labeling

In [ ]:
def create_mood_label(row):
    if row['joy'] > 0.5:
        return 'Happy'
    elif row['sadness'] > 0.5:
        return 'Sad'
    elif row['anger'] > 0.5:
        return 'Angry'
    # Add more rules as needed
    else:
        return 'Neutral'

data_with_emotions['mood_label'] = data_with_emotions.apply(create_mood_label, axis=1)


### Clustered Labels

In [ ]:
from sklearn.metrics import silhouette_score

# Assuming 'features' is the matrix of features we're using for clustering
features = data_with_emotions[['anger', 'disgust', 'fear', 'joy', 'neutral', 'sadness', 'surprise']].values

# We'll test silhouette scores for different numbers of clusters
range_n_clusters = list(range(2, 11))  # For example, testing 2 to 10 clusters
silhouette_avg_scores = []

for n_clusters in range_n_clusters:
    clusterer = KMeans(n_clusters=n_clusters, random_state=42)
    cluster_labels = clusterer.fit_predict(features)
    silhouette_avg = silhouette_score(features, cluster_labels)
    silhouette_avg_scores.append(silhouette_avg)
    print(f"For n_clusters = {n_clusters}, the average silhouette_score is : {silhouette_avg}")

# Plot the silhouette scores
plt.figure(figsize=(10, 6))
plt.plot(range_n_clusters, silhouette_avg_scores, marker='o')
plt.title('Silhouette Analysis For Optimal k')
plt.xlabel('Number of clusters (k)')
plt.ylabel('Silhouette Score')
plt.show()


In [ ]:
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt

# Assuming data_with_emotions is your DataFrame with the emotion scores as separate columns
# We will use these scores as features for clustering
features = data_with_emotions[['anger', 'disgust', 'fear', 'joy', 'neutral', 'sadness', 'surprise']].values

# Choose the number of clusters (k) you want to divide your data into
# The choice of k could be informed by domain knowledge, or you could use methods like the elbow method to choose k
k = 3

# Perform K-Means clustering
kmeans = KMeans(n_clusters=k, random_state=42)
data_with_emotions['cluster'] = kmeans.fit_predict(features)

# Now let's see the clusters
# Since we have multiple dimensions (each emotion score is a dimension), let's reduce it to two dimensions using PCA
from sklearn.decomposition import PCA
pca = PCA(n_components=2)
reduced_features = pca.fit_transform(features)

# Plot the reduced features in a scatter plot and color them with their cluster label
plt.scatter(reduced_features[:, 0], reduced_features[:, 1], c=data_with_emotions['cluster'], cmap='viridis')
plt.xlabel('PCA Feature 1')
plt.ylabel('PCA Feature 2')
plt.title('Anime Emotion Scores Clustering')
plt.colorbar(label='Cluster Label')

# Show the centroids of each cluster in the reduced feature space
centroids = pca.transform(kmeans.cluster_centers_)
plt.scatter(centroids[:, 0], centroids[:, 1], marker='X', s=200, color='r')

plt.show()


### Identifying Super-users

Using master_data

In [19]:
# Calculate the number of interactions per user
interaction_counts = master_data.groupby('profile').size()

# Using percentiles to define super-users
percentile_threshold = 95  # Top 5% of users
threshold = interaction_counts.quantile(percentile_threshold / 100)
super_users = interaction_counts[interaction_counts > threshold].index.tolist()

print(f"Threshold for super-users (top {100 - percentile_threshold}%): {threshold:.2f}")
print(f"Identified {len(super_users)} super-users.")


Threshold for super-users (top 5%): 56.00
Identified 970 super-users.


Using data_with_emotions subset - approx. 34k

In [21]:
# Calculate the number of interactions per user
interaction_counts = data_with_emotions.groupby('profile').size()

# Using percentiles to define super-users
percentile_threshold = 95  # Top 5% of users
threshold = interaction_counts.quantile(percentile_threshold / 100)
super_users = interaction_counts[interaction_counts > threshold].index.tolist()

print(f"Threshold for super-users (top {100 - percentile_threshold}%): {threshold:.2f}")
print(f"Identified {len(super_users)} super-users.")


Threshold for super-users (top 5%): 10.00
Identified 497 super-users.


### New dataset with and without super-users

In [ ]:
#super_users = 

### Clustering-Based Hybrid Methods
Perform clustering on user and item features and then apply collaborative and content-based methods within each cluster.

### Ensemble Methods
Combine multiple different recommendation models using ensemble techniques. For example, using a voting or stacking approach where predictions from multiple models are combined to produce the final output.

### Neural Collaborative Filtering (NCF)
NCF models use neural networks to learn user-item interaction patterns, providing a more flexible and powerful way to model complex relationships compared to traditional matrix factorization methods.

In [22]:
pip install tensorflow

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 208.8 MB 8.1 kB/s  eta 0:00:01████████████████▍             | 119.6 MB 79.2 MB/s eta 0:00:02
     |████████████████████████████████| 394 kB 72.3 MB/s eta 0:00:01
  Using cached flatbuffers-23.5.26-py2.py3-none-any.whl (26 kB)
     |████████████████████████████████| 1.2 MB 44.7 MB/s eta 0:00:01
  Using cached tensorflow_estimator-2.15.0-py2.py3-none-any.whl (441 kB)
  Using cached termcolor-2.4.0-py3-none-any.whl (7.7 kB)
     |████████████████████████████████| 1.9 MB 39.6 MB/s eta 0:00:01
     |████████████████████████████████| 2.7 MB 80.8 MB/s eta 0:00:01
  Using cached opt_einsum-3.3.0-py3-none-any.whl (65 kB)
  Using cached google_pasta-0.2.0-py3-none-any.whl (57 kB)
     |████████████████████████████████| 20.6 MB 46.9 MB/s eta 0:00:01
  Using cached absl_py-2.0.0-py3-none-any.whl (130 kB)
     |████████████████████████████████| 9.6 MB 66.6 MB/s eta 0:00:01
  Using c

In [23]:
pip install keras

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


In [24]:
from keras.models import Model
from keras.layers import Embedding, Flatten, Input, concatenate, Dense
from keras.optimizers import Adam
from sklearn.model_selection import train_test_split
from tensorflow import keras

# Columns: 'uid_review' for users, 'anime_uid' for items, and 'Overall' for ratings
# Convert user and item IDs to categorical variables
data_with_emotions['uid_review'] = data_with_emotions['uid_review'].astype('category')
data_with_emotions['anime_uid'] = data_with_emotions['anime_uid'].astype('category')

# Map user and item IDs to integers
user_id_mapping = {id: i for i, id in enumerate(data_with_emotions['uid_review'].cat.categories)}
item_id_mapping = {id: i for i, id in enumerate(data_with_emotions['anime_uid'].cat.categories)}

# Apply mapping
data_with_emotions['user'] = data_with_emotions['uid_review'].map(user_id_mapping)
data_with_emotions['item'] = data_with_emotions['anime_uid'].map(item_id_mapping)

In [25]:
# Prepare the data for the model
X = data_with_emotions[['user', 'item']].values
y = data_with_emotions['Overall'].values

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=13)

# Neural Collaborative Filtering model architecture
user_input = Input(shape=(1,))
item_input = Input(shape=(1,))
user_embedding = Embedding(len(user_id_mapping), 15, input_length=1)(user_input)
item_embedding = Embedding(len(item_id_mapping), 15, input_length=1)(item_input)
user_vec = Flatten()(user_embedding)
item_vec = Flatten()(item_embedding)
concat = concatenate([user_vec, item_vec])
dense = Dense(64, activation='relu')(concat)
output = Dense(1)(dense)
model = Model([user_input, item_input], output)

# Compile and train the model
model.compile(optimizer=Adam(0.001), loss='mean_squared_error')
model.fit([X_train[:, 0], X_train[:, 1]], y_train, batch_size=64, epochs=5, validation_split=0.1)

# Evaluate the model
model.evaluate([X_test[:, 0], X_test[:, 1]], y_test)

Epoch 1/5
394/394 [==============================] - 1s 1ms/step - loss: 22.6378 - val_loss: 4.1832
Epoch 2/5
394/394 [==============================] - 1s 1ms/step - loss: 2.2857 - val_loss: 2.8225
Epoch 3/5
394/394 [==============================] - 1s 1ms/step - loss: 0.6315 - val_loss: 2.7547
Epoch 4/5
394/394 [==============================] - 1s 1ms/step - loss: 0.3145 - val_loss: 2.7346
Epoch 5/5
219/219 [==============================] - 0s 289us/step - loss: 2.6124


2.6123688220977783

## Interpretation

The model is a Neural Collaborative Filtering (NCF) model, which is designed to predict user ratings for items (anime) based on learned user and item embeddings.

### Model Architecture:

- Model has separate embedding layers for users (user_embedding) and items (item_embedding), each with an embedding size of 15. This means the model learns a 15-dimensional vector to represent each user and each item.

- Embeddings are flattened and then concatenated.

- After concatenation, the data passes through a dense layer with 64 neurons and ReLU activation, which allows the model to learn non-linear relationships.

- The final output layer has 1 neuron (since this is a regression problem where you're predicting a rating value).

### Compilation and Training:

- The model is compiled with the Adam optimizer and mean squared error (MSE) loss.

- The model is trained for 5 epochs with a batch size of 64 and uses 10% of the training data for validation.


### Model Evaluation:

- After training, the model is evaluated on the test set, resulting in a MSE of approximately 2.61.

- The MSE is a measure of the average squared difference between the actual ratings and the ratings predicted by the model. A lower MSE indicates better performance.

- This MSE value means that on average, the predicted rating deviates from the actual rating by the square root of 2.612, which is around 1.62. 


### Interpreting the Results:

- The scale of ratings is 1 to 10, an average deviation of around 1.6 might be considered moderate. It shows that the model has learned to some extent but still has room for improvement.


### Next Steps for Improvement:

- Consider training the model for more epochs or tuning other hyperparameters.
- Experiment with different architectures, like adding more dense layers or changing the number of neurons.
- Explore advanced techniques like regularization (to prevent overfitting) and learning rate scheduling.
- Consider content-based features (like genres or synopses), consider a hybrid model that combines collaborative and content-based approaches.


## Hybrid Deep Learning Models
Combine features from both content-based and collaborative filtering in a deep learning architecture, allowing the model to learn complex interactions between content and user preferences.

### Content-Based Filtering
Feature Extraction from Synopsis using TF-IDF

In [27]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Fill NaN values in 'synopsis' with empty string
data_with_emotions['synopsis'].fillna('', inplace=True)

# Apply TF-IDF
tfidf = TfidfVectorizer(stop_words='english')
content_features = tfidf.fit_transform(data_with_emotions['synopsis'])


### Prepare Collaborative Filtering Data
Prepare user-item interaction data and create embeddings for collaborative filtering

In [28]:
from keras.layers import Embedding, Input, Flatten

# Prepare user and item interaction data ('uid_review' and 'anime_uid' as user and item identifiers)
user_ids = data_with_emotions['uid_review'].astype('category').cat.codes.values
item_ids = data_with_emotions['anime_uid'].astype('category').cat.codes.values

num_users = data_with_emotions['uid_review'].nunique()
num_items = data_with_emotions['anime_uid'].nunique()
embedding_size = 20

# User embedding
user_input = Input(shape=(1,))
user_embedding = Embedding(num_users, embedding_size, input_length=1)(user_input)
user_vec = Flatten()(user_embedding)

# Item embedding
item_input = Input(shape=(1,))
item_embedding = Embedding(num_items, embedding_size, input_length=1)(item_input)
item_vec = Flatten()(item_embedding)


### Hybrid Model Architecture
Combine content-based and collaborative filtering features in a neural network

In [ ]:
from keras.layers import concatenate, Dense
from keras.models import Model

# Combine user_vec and item_vec with content features
# Assume content_features is a dense representation of content
concatenated = concatenate([user_vec, item_vec, content_features])

# Add dense layers for deep learning
dense = Dense(128, activation='relu')(concatenated)
output = Dense(1, activation='linear')(dense)


model = Model(inputs=[user_input, item_input, content_input], outputs=output)

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')


### Training the Model
Train the model using your dataset. Note that the dataset needs to be split into training and testing sets, and the content features should be in a compatible format for training.

In [ ]:
# Split data


In [ ]:
# Train the model
model.fit([user_ids, item_ids, content_features], small_sample['rating'], epochs=10, batch_size=32)

### Mood-Based Classification
Tagging Anime with Moods from Emotion Classifier

In [ ]:
# Assuming mood classification has already been done and each anime has a 'mood' column
mood_based_data = small_sample[['anime_id', 'mood']]


### Combining Recommendations Models
Combining collaborative and content-based filtering


In [ ]:
def hybrid_recommendation(user_id, content_rec, collaborative_rec, mood=None):
    # Hybrid model combining content and collaborative filtering
    hybrid_rec = content_rec * 0.5 + collaborative_rec * 0.5

    # If mood is specified, filter or re-rank recommendations
    if mood:
        hybrid_rec = filter_or_rank_by_mood(hybrid_rec, mood)

    return hybrid_rec


In [ ]:
def hybrid_recommendation(user_id, mood=None):
    # Get content-based recommendations
    content_based_rec = content_based_model(user_id)

    # Get collaborative filtering recommendations
    collaborative_rec = collaborative_model(user_id)

    # Get mood-based recommendations if a mood is specified
    if mood:
        mood_based_rec = mood_based_model(user_id, mood)
    else:
        mood_based_rec = {}

    # Combine recommendations
    # This is a simple average. Consider using more complex methods
    combined_rec = average(content_based_rec, collaborative_rec, mood_based_rec)

    return combined_rec


### Personalization and Playlist Generation

In [ ]:
def generate_playlist(user_id, mood=None):
    content_rec = get_content_based_recommendations(user_id)
    collaborative_rec = get_collaborative_recommendations(user_id)
    
    # Hybrid recommendations
    recommendations = hybrid_recommendation(user_id, content_rec, collaborative_rec, mood)

    # Generate playlist based on recommendations
    playlist = create_playlist(recommendations)
    return playlist


### Model Evaluation and Tuning

In [ ]:
from sklearn.metrics import precision_score

# Assuming you have true_labels and predicted_labels
precision = precision_score(true_labels, predicted_labels)
